src/planner.c

```diff
@@ -528,29 +505,59 @@ timescaledb_set_rel_pathlist(PlannerInfo *root,
 							 RangeTblEntry *rte)
 {
 	Hypertable *ht;
-
-	global_planner_context_init();
+	Cache	   *hcache;
 
 	if (prev_set_rel_pathlist_hook != NULL)
 		(*prev_set_rel_pathlist_hook) (root, rel, rti, rte);
 
-	if (!extension_is_loaded() ||
-		IS_DUMMY_REL(rel) ||
-		!should_optimize_query() ||
-		!OidIsValid(rte->relid))
-		goto out_release;
+	if (!extension_is_loaded() || IS_DUMMY_REL(rel) || !OidIsValid(rte->relid))
+		return;
 
-	sort_transform_optimization(root, rel);
+	hcache = hypertable_cache_pin();
+	ht = hypertable_cache_get_entry(hcache, rte->relid);
 
-	if (rel->rtekind != RTE_RELATION ||
-		rte->relkind != RELKIND_RELATION ||
-	/* Do not optimize result relations (INSERT, UPDATE, DELETE) */
-		rti == (Index) root->parse->resultRelation)
+	if (!should_optimize_query(ht))
 		goto out_release;
 
-	ht = hypertable_cache_get_entry(gpc.hcache, rte->relid);
+	if (guc_optimize_non_hypertables)
+	{
+		/* if optimizing all tables, apply optimization to any table */
+		sort_transform_optimization(root, rel);
+	}
+	else if (ht != NULL && rel->reloptkind == RELOPT_OTHER_MEMBER_REL)
```

Comment:
Seems here you should first check for `RELOPT_OTHER_MEMBER_REL` then do the hypertable lookup rather than the other way around. Although I see that the hypertable is also used for the append optimization. Is there a way to make this cleaner and only lookup the hypertable if needed?